File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: Optional[ListNode]
        :type n: int
        :rtype: Optional[ListNode]
        """

        # Two pointers approach (one pass)

        # Dummy pointer before head;
        dummy = ListNode(0)
        dummy.next = head

        # Place the first pointer n nodes ahead
        first = head
        for i in range(n):
            first = first.next

        # Start second pointer from start and first from nth node, and go ahead one by one 
        second = dummy
        while first:
            second = second.next
            first = first.next

        second.next = second.next.next

        return dummy.next 

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let length = 0;
    let temp = head;

    // Step 1: Find length
    while (temp) {
        length++;
        temp = temp.next;
    }

    // Step 2: Find the position to remove (from start)
    let prevPosition = length - n;

    // Step 3: Use a dummy node
    let dummy = new ListNode(0);
    dummy.next = head;
    let previous = dummy;

    // Step 4: Move to one previous node
    for (let i = 0; i < prevPosition; i++) {
        previous = previous.next;
    }

    // Step 5: Remove node
    previous.next = previous.next.next;

    // Step 6: Return new head
    return dummy.next;
};
